src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,207 @@
+    private static final String ID_TAG = "@id";
+
+    private final String myExceptionPage;
+
+    private final String myFesterizeVersion;
+
+    /**
+     * FUA = Festerize User Agent
+     */
+    private final Pattern myFUAPattern;
+
+    /**
+     * @param aVertx
+     * @param aConfig
+    */
+    public PostThumbnailsHandler(final Vertx aVertx, final JsonObject aConfig) throws IOException {
+        super(aVertx, aConfig);
+
+        final byte[] bytes = IOUtils.readBytes(getClass().getResourceAsStream("/webroot/error.html"));
+
+        myExceptionPage = new String(bytes, StandardCharsets.UTF_8);
+        myFesterizeVersion = aConfig.getString(Config.FESTERIZE_VERSION);
+        myFUAPattern = Pattern.compile("Festerize/(?<version>\\d+\\.\\d+\\.\\d+)");
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see io.vertx.core.Handler#handle(java.lang.Object)
+     */
+    @Override
+    public void handle(final RoutingContext aContext) {
+        final HttpServerRequest request = aContext.request();
+        final HttpServerResponse response = aContext.response();
+        final Set<FileUpload> csvUploads = aContext.fileUploads();
+        final String festerizeUserAgent = StringUtils.trimTo(request.getHeader("User-Agent"), Constants.EMPTY);
+        final Matcher festerizeUserAgentMatcher = myFUAPattern.matcher(festerizeUserAgent);
+        final String errorMessage;
+
+        // An uploaded CSV is required
+        if (csvUploads.isEmpty()) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_037);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else if (festerizeUserAgentMatcher.matches() &&
+                   !festerizeUserAgentMatcher.group("version").equals(myFesterizeVersion)) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_147, myFesterizeVersion);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else {
+            final FileUpload csvFile = csvUploads.iterator().next();
+            final String filePath = csvFile.uploadedFileName();
+            final String fileName = csvFile.fileName();
+
+            try {
+                final List<String[]> originalLines =
+                    new CSVReader(Files.newBufferedReader(Paths.get(filePath))).readAll();
+                final List<String[]> linesWithThumbs = ThumbnailUtils.addThumbnailColumn(originalLines);
+                final int manifestIndex = Arrays.asList(linesWithThumbs.get(0)).indexOf(CSV.MANIFEST_URL);
+                final String collectionURL = linesWithThumbs.get(1)[manifestIndex];
+                final Future<JsonObject> collection = getManifest(collectionURL);
+                collection.onComplete(result -> {
+                    if (result.failed()) {
+                        returnError(response, HTTP.INTERNAL_SERVER_ERROR, result.cause().getMessage());
+                    } else {
+                        final JsonArray canvases = result.result().getJsonArray(MANIFESTS);
```

Comment:
I'm a little confused by the variable name here. You're requesting the collection document (https://iiif.library.ucla.edu/collections/ark%3A%2F21198%2Fzz001ng4t6) and then getting the child work manifests of that collection from the returned JSON. I'm not sure why they're called canvases here.

If you pull each of the child manifests from your collection doc, you'll see the canvases in their JSON (e.g., https://iiif.library.ucla.edu/ark%3A%2F21198%2Fzz001p6m70/manifest). So, I think down below your `canvasURL` is really the `manifestURL`. You're then doing a `getManifest(URL)` on them and returning the work manifest.

Each of those work manifests (JSON) will have multiple canvases and I believe you'd want to pick the representative from those canvases for the work they're contained in (whereas now I think you'd random sampling the work manifests on the collection doc). You'd then need a way to associate the work you've pulled from the collection down with the work that's on the CSV and add its thumbnail base URL (it's ID, I'd assume).

If I'm misreading things, let me know. We can chat (or video) in Slack (which might be easier than commenting on here?)